src/proof/lfsc/lfsc_util.h

```diff
@@ -0,0 +1,105 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds, Yoni Zohar
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Utilities for LFSC proofs.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC4__PROOF__LFSC__LFSC_UTIL_H
+#define CVC4__PROOF__LFSC__LFSC_UTIL_H
+
+#include <iostream>
+#include <map>
+
+#include "expr/node.h"
+#include "proof/proof_letify.h"
+
+namespace cvc5 {
+namespace proof {
+
+/**
+ * LFSC rules. The enum below contains all rules that don't correspond to a
+ * PfRule, e.g. congruence in LFSC does not have the same form as congruence
+ * in the internal calculus.
+ */
+enum class LfscRule : uint32_t
+{
+  //----------- translated rules
+
+  // We defined LFSC versions for rules that either don't exist in the internal
+  // calculus, or have a different set of arugments/children.
+
+  // scope has a different structure, e.g. uses lambdas
+  SCOPE,
+  // must distinguish equalities and disequalities
+  NEG_SYMM,
+  // congruence is done via a higher-order variant of congruence
+  CONG,
+  // we use unrolled binary versions of and intro
+  AND_INTRO1,
+  AND_INTRO2,
+  // needed as a helper for SCOPE
+  NOT_AND_REV,
+  PROCESS_SCOPE,
+  // arithmetic
+  ARITH_SUM_UB,
+
+  // form of quantifier rules varies from internal calculus
+  INSTANTIATE,
+  SKOLEMIZE,
+
+  // a lambda with argument
+  LAMBDA,
+  // a proof-let "plet"
+  PLET,
+  //----------- unknown
+  UNKNOWN,
+};
+
+/**
+ * Converts a lfsc rule to a string. Note: This function is also used in
+ * `safe_print()`. Changing this function name or signature will result in
```

Comment:
say where this method is

Reply:
Which method? Note this is the standard documentation for all methods of this form.

Reply:
`safe_print()`, I meant. But ok, I had never noticed the methods of this form nor knew about this safe_print stuff.